File: programmers/Level1/solution/_46_성격_유형_검사하기/Solution.java
package programmers.Level1.solution._46_성격_유형_검사하기;

import java.util.HashMap;

/**
 * https://school.programmers.co.kr/learn/courses/30/lessons/118666?language=java
 */
public class Solution {

    public static void main(String[] args) {

        String[] survey = { "AN", "CF", "MJ", "RT", "NA" };
        int[] choices = { 5, 3, 2, 7, 5 };

        Solution solution = new Solution();
        String result = solution.solution(survey, choices);
        System.out.println(result);

    }

    public String solution(String[] survey, int[] choices) {

        char[][] surveyType = { { 'R', 'T' }, { 'C', 'F' }, { 'J', 'M' }, { 'A', 'N' } };

        HashMap<Character, Integer> userMap = new HashMap<>();
        for (int i = 0; i < survey.length; i++) {
            int choice = choices[i];
            // 4점 미만인 경우 앞의 사람 점수 증가
            if (choice < 4) {
                char user = survey[i].charAt(0);
                userMap.put(user, userMap.getOrDefault(user, 0) + 4 - choice);

            } else if (4 < choice) { // 4점 초과인 경우 뒷 사람 점수가 증가
                char user = survey[i].charAt(1);
                userMap.put(user, userMap.getOrDefault(user, 0) + choice - 4);
            }
        }

        // 유저 타입에 따라 성격 유형 판단
        StringBuilder result = new StringBuilder();
        for (char[] type : surveyType) {
            int leftScore = userMap.getOrDefault(type[0], 0);
            int rightScore = userMap.getOrDefault(type[1], 0);
            if (rightScore <= leftScore) {
                result.append(type[0]);
            } else {
                result.append(type[1]);
            }
        }

        return result.toString();
    }

}

File: programmers/Level1/solution/_46_성격_유형_검사하기/Solution.kt
/*
 * https://school.programmers.co.kr/learn/courses/30/lessons/118666?language=kotlin
 */
fun main() {
    var survey: Array<String> = arrayOf("AN", "CF", "MJ", "RT", "NA")
    var choices: IntArray = intArrayOf(5, 3, 2, 7, 5)

    var result = solution(survey, choices)
    print(result)
}

fun solution(survey: Array<String>, choices: IntArray): String {
    var surveyType: Array<CharArray> =
        arrayOf(
            charArrayOf('R', 'T'),
            charArrayOf('C', 'F'),
            charArrayOf('J', 'M'),
            charArrayOf('A', 'N')
        )

    var userMap: HashMap<Char, Int> = HashMap()
    for (i in survey.indices) {
        var choice: Int = choices[i]
        if (choice < 4) {
            var user: Char = survey[i].get(0)
            userMap.put(user, userMap.getOrDefault(user, 0) + 4 - choice)
        } else if (4 < choice) {
            var user: Char = survey[i].get(1)
            userMap.put(user, userMap.getOrDefault(user, 0) + choice - 4)
        }
    }

    var result: StringBuilder = StringBuilder()
    (surveyType).forEach { type ->
        var leftScore: Int = userMap.getOrDefault(type[0], 0)
        var rightScore: Int = userMap.getOrDefault(type[1], 0)
        if (rightScore <= leftScore) {
            result.append(type[0])
        } else {
            result.append(type[1])
        }
    }
    return result.toString()
}
